openwhisk/executor.go

```diff
@@ -0,0 +1,208 @@
+		return nil
+	}
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil
+	}
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil
+	}
+
+	pipeOut, pipeIn, err := os.Pipe()
+	if err != nil {
+		return nil
+	}
+	cmd.ExtraFiles = []*os.File{pipeIn}
+
+	return &Executor{
+		make(chan string),
+		make(chan bool),
+		make(chan error),
+		cmd,
+		bufio.NewWriter(stdin),
+		bufio.NewScanner(pipeOut),
+		bufio.NewScanner(stdout),
+		bufio.NewScanner(stderr),
+	}
+}
+
+// collect log from a stream
+func _collect(ch chan string, scan *bufio.Scanner) {
+	for scan.Scan() {
+		ch <- scan.Text()
+	}
+}
+
+// loop over the command executing
+// returning when the command exits
+func (proc *Executor) run() {
+	log.Println("run: start")
+	err := proc._cmd.Start()
+	if err != nil {
+		proc.exit <- err
+		log.Println("run: early exit")
+		proc._cmd = nil // do not kill
+		return
+	}
+	// wait for the exit
+	proc.exit <- proc._cmd.Wait()
+	proc._cmd = nil // do not kill
+	log.Println("run: end")
+}
+
+// manage copying stdout and stder in output
+// with log guards
+func (proc *Executor) logger() {
+	log.Println("logger: start")
+	// poll stdout and stderr
+	chOut := make(chan string)
+	go _collect(chOut, proc._logout)
+	chErr := make(chan string)
+	go _collect(chErr, proc._logerr)
+
+	// wait for the signal
+	for <-proc.log {
+		// flush stdout
+		runtime.Gosched()
+		for loop := true; loop; {
+			select {
+			case buf := <-chOut:
+				fmt.Println(buf)
+			case <-time.After(TIMEOUT):
+				loop = false
+			}
+		}
+		fmt.Println("XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX")
+
+		// flush stderr
+		runtime.Gosched()
+		for loop := true; loop; {
+			select {
+			case buf := <-chErr:
+				fmt.Println(buf)
+			case <-time.After(TIMEOUT):
+				loop = false
+			}
+		}
+		fmt.Println("XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX")
+	}
+	log.Printf("logger: end")
+}
+
+// main service function
+// writing in input
+// and reading in output
+// using the provide channels
+func (proc *Executor) service() {
+	log.Println("service: start")
+	for {
+		in := <-proc.io
+		if in == "" {
+			log.Println("terminated upon request")
+			break
+		}
+		// input/output with the process
+		log.Printf(">>>%s\n", in)
+		proc._input.WriteString(in + "\n")
+		proc._input.Flush()
+		if proc._output.Scan() {
+			out := proc._output.Text()
+			log.Printf("<<<%s\n", out)
+			proc.io <- out
+			if out == "" {
+				break
+			}
+		}
+	}
+	log.Printf("service: end")
+}
+
+// Start execution of the command
+// returns an error if the program fails
+func (proc *Executor) Start() error {
+	// start the underlying executable
+	// check if died
+	go proc.run()
+	select {
+	case <-proc.exit:
+		// oops, it died
+		return fmt.Errorf("command exited")
+	case <-time.After(TIMEOUT):
+		// ok let's process it
+		go proc.service()
+		go proc.logger()
+	}
+	return nil
+}
+
+// Stop will kill the process
+// and close the channels
+func (proc *Executor) Stop() {
+	log.Println("stopping")
```

Comment:
this is another method that for now won't be called with the current container lifecycle; a comment would be nice to that effect if you agree.

in general there are a few flows that can be tested and provide a clean interface but should be qualified to separate the behavior against the invoker.

Reply:
Yup I can refactor, I will make you happy. It is just sometimes my coding style is a bit... lazy :)

Reply:
hehe - no worries you can leave it. I was able to follow the code nicely even though I’ve written a grand total of 100 lines of go code.